### PR TITLE
ACF connector test implemented and ACF v5.x support added

### LINKED ISSUE #1122: Support for Advanced Custom Fields (ACF) version 5
> When updating an ACF repeater field (which is inside an ACF Group field),
the Stream plugin crashes as the mysql escape function from WP_DB gets incorrect parameters passed.

https://wordpress.org/support/topic/error-crash-when-updating-acf-repeater-field/

### PR DESCRIPTION
**Pull request by kidunot89**
Partially fixes #1093 
Fixes and closes #1122 

# Summary
- Updates the ACF connector class to work with the changes made in ACF v5+
- Adds tests for ACF v5+ connector logic.
- Adds `wpackagist` repository and `wpackagist-plugin/advanced-custom-fields` dev dependency to the `composer.json`.
- Updates tests `bootstrap` file to activate plugins defined in the PHPUnit const `WP_TEST_ACTIVATED_PLUGINS`

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -19,6 +23,7 @@
     "wordpress/wordpress": "^5.2",
     "wp-cli/wp-cli-bundle": "^2.2",
     "wp-coding-standards/wpcs": "^2.2",
+    "wpackagist-plugin/advanced-custom-fields": "*",
```
Comment (kopepasah) [inline composer.json:26]: @kidunot89 forgive my popping in without context, but what is the reason for requiring ACF? This believe will install this plugin in cases someone manages Stream installations using Composer.

Comment (kidunot89) [reply composer.json:26]: :thinking: See as this is a **dev** dependency, I don't believe that is the case.

Comment (kopepasah) [reply composer.json:26]: @kidunot89 whoops. I did not expand the view and mistakenly thought this was under `require`. 😬  I will not install as require-dev is read a `root` only.

File: connectors/class-connector-acf.php [hunk 2]
```diff
@@ -147,13 +149,219 @@ public function action_links( $links, $record ) {
...
+	}
+
 	/**
 	 * Track addition of post meta
 	 *
 	 * @action added_post_meta
 	 */
 	public function callback_added_post_meta() {
-		call_user_func_array( array( $this, 'check_meta' ), array_merge( array( 'post', 'added' ), func_get_args() ) );
+		$this->check_meta( 'post', 'added', ...func_get_args() );
```
Comment (kopepasah) [inline connectors/class-connector-acf.php:364]: @kidunot89 what is the PHP version compatibility of the spread operator?

**Comment by kopepasah** [comment]
@kidunot89 my only concern here is requiring the ACF plugin from Composer for development. If we were to follow this same practice, we could end up with a ton of WordPress plugins we are adding the the Composer config, which are not really required for development, but rather for testing. Installing of these plugins for the PHPUnit tests may better be served in some other scripted process for testing, but would like to hear your thoughts on why this method is prefered.

**Comment by kidunot89** [comment]
@kopepasah Yes, My plan is to add all the plugins, supported by stream through connectors, to the `dev` dependencies, and writing an integration test for each connector. My reasoning is simple, is the best option for testing the **connectors**.

File: connectors/class-connector-acf.php (hunk 2 above)
Comment (kidunot89) [reply connectors/class-connector-acf.php:364]: :man_facepalming: 7.4, my bad :sweat_smile:.

Comment (kidunot89) [reply connectors/class-connector-acf.php:364]: @kopepasah Upon doing further research, I learned that this is called **Argument Unpacking** in PHP, and yes it was implemented in PHP 5.6. Official RFC [here](https://wiki.php.net/rfc/argument_unpacking)

**Comment by kasparsd** [comment]
@kidunot89 I suspect the failing unit tests are caused by wp-dev-lib copying over the project source files to a different place for the actual test run:

https://github.com/xwp/wp-dev-lib/blob/420f1bac828e105a19a904cab97d042891d4b84e/scripts/check-diff.sh#L757-L765

So none of the dependencies (plugins) are moved over to the same `wp-content/plugins` directory. This is a great example of the many assumptions wp-dev-lib makes about the project structure.

I guess at this point we can switch to using the Docker test environment on Travis too. There are ways to cache the Docker images once they're built so we should be fine.

File: connectors/class-connector-acf.php (hunk 2 above)
Comment (kopepasah) [reply connectors/class-connector-acf.php:364]: Thanks for the additional research! I was not aware it had been around for so long. 😱

**Comment by kidunot89** [comment]
@kasparsd I don't believe that to be the case due to the fact that the same configurations are being used in the bbPress connector PR and CI is passing :point_right: #1120

**Comment by kasparsd** [comment]
@kidunot89 Unfortunately this is another issue with wp-dev-lib -- those tests are passing because it didn't even run the tests https://travis-ci.com/github/xwp/stream/jobs/355627134#L490 -- the `source "$DEV_LIB_PATH/travis.script.sh"` command failed silently...

**Comment by kopepasah** [comment]
@kidunot89 just a quick question, regarding ACF v5. Do we need to perform any version checks for backwards compatibility or does the functionality simply not exist pre-v5?

**Comment by kidunot89** [comment]
@kopepasah Pre-v5 functionality exists, it's just not tested, or updated.

File: connectors/class-connector-acf.php [hunk 3]
```diff
@@ -147,13 +149,210 @@ public function action_links( $links, $record ) {
...
+		// Log new ACF field additions to field groups.
+		if ( 'acf-field' === $post->post_type ) {
+			$parent = get_post( $post->post_parent );
+			if ( $parent ) {
+				$this->log_prop( 'added', $post_id, $post, 'parent', $parent );
+			}
+		} elseif ( 'acf-field-group' === $post->post_type ) {
+			$props = maybe_unserialize( $post->post_content );
+
+			if ( ! empty( $props ) ) {
```
Comment (kopepasah) [inline connectors/class-connector-acf.php:176]: @kidunot89 `maybe_unserialize()` does not always return an array. Do you think we need to check that `$props` is an array before running the `foreach` loop?

File: connectors/class-connector-acf.php [hunk 4]
```diff
@@ -479,10 +678,31 @@ public function log_override( $data ) {
 			return $data;
 		}
 
-		if ( 'posts' === $data['connector'] && 'acf' === $data['context'] ) {
-			$data['context']               = 'field_groups';
-			$data['connector']             = $this->name;
-			$data['args']['singular_name'] = esc_html__( 'field group', 'stream' );
+		$is_acf_context = 'acf' === $data['context']
```
Comment (kopepasah) [inline connectors/class-connector-acf.php:681]: @kidunot89 how about this: 

```php
$is_acf_context = in_array( $data['context'], array( 'acf', 'acf-field-group', 'acf-field' ), true );
```

File: tests/tests/connectors/test-class-connector-acf.php [hunk 5]
```diff
@@ -0,0 +1,193 @@
+<?php
+namespace WP_Stream;
```
Comment (kopepasah) [inline tests/tests/connectors/test-class-connector-acf.php:2]: I am not certain how @kasparsd feels, but I would like to see some more documentation in these tests.

File: tests/tests/connectors/test-class-connector-acf.php [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+		$this->mock = $this->getMockBuilder( Connector_ACF::class )
+			->setMethods( [ 'log' ] )
+			->getMock();
+
+		// Register connector.
+		$this->mock->register();
+
+		// Reusable ACF object keys.
+		$this->group_key = __CLASS__;
+		$this->field_key = 'field_5d7812fd000a4';
```
Comment (kopepasah) [inline tests/tests/connectors/test-class-connector-acf.php:22]: @kidunot89 is this field key unique (always expected to be this value) or arbitrary?

File: tests/tests/connectors/test-class-connector-acf.php [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+		$defaults = [
+			'parent'            => $this->group_key,
+			'key'               => $this->field_key,
+			'label'             => 'Test Field',
+			'name'              => 'test_field',
+			'type'              => 'text',
+			'instructions'      => '',
+			'required'          => 0,
+			'conditional_logic' => 0,
+			'wrapper'           => array(
```
Comment (kopepasah) [inline tests/tests/connectors/test-class-connector-acf.php:66]: Sometimes you are using shorthand arrays and other times using the long form of array. We need to pick one an stick with that form throughout the plugin.

**Comment by kopepasah** [comment]
> @kopepasah Pre-v5 functionality exists, it's just not tested, or updated.

Thanks for that information! It is difficult to tell, just by reviewing, which parts are pre-v5 and remain backwards compatible, but as long as you are aware, then that can be resolved when testing the changes.

I added my review and requested just a few changes.

File: connectors/class-connector-acf.php (hunk 4 above)
Comment (kidunot89) [reply connectors/class-connector-acf.php:681]: :man_facepalming: Can't believe I didn't think of that :joy:

File: tests/tests/connectors/test-class-connector-acf.php (hunk 5 above)
Comment (kidunot89) [reply tests/tests/connectors/test-class-connector-acf.php:2]: There was a doc block, but it was lost in PR somewhere. I'll correct this.

File: tests/tests/connectors/test-class-connector-acf.php (hunk 6 above)
Comment (kidunot89) [reply tests/tests/connectors/test-class-connector-acf.php:22]: No, I was using this in an old iteration the test, I remove it.

File: tests/tests/connectors/test-class-connector-acf.php (hunk 7 above)
Comment (kidunot89) [reply tests/tests/connectors/test-class-connector-acf.php:66]: For now, I go with long version because it used through the plugin.

**Comment by kidunot89** [comment]
@kopepasah I add notes in the doc blocks for the pre ACF-v5 callbacks.

File: tests/tests/connectors/test-class-connector-acf.php (hunk 5 above)
Comment (kidunot89) [reply tests/tests/connectors/test-class-connector-acf.php:7]: This has been [fixed](https://github.com/xwp/stream/pull/1118/files#diff-8b4d036afce8fdd07963c1eba24744a1R2-R6), but the snippet isn't updated :thinking:

**Comment by kasparsd** [comment]
This looks good @kidunot89! Feel free to merge if there is nothing to add.

**Pull request merged**